k8s/envoy-jwt-opa/README.md

```diff
@@ -0,0 +1,336 @@
+            "content-length": "0",
+            "user-agent": "Go-http-client/1.1",
+            "x-forwarded-proto": "http",
+            "x-request-id": "fad45df6-3cc1-4ce9-9cad-fb3b65eff037"
+          },
+          "host": "localhost:3001",
+          "id": "10476077497628160603",
+          "method": "GET",
+          "path": "/transactions/1",
+          "protocol": "HTTP/1.1"
+        },
+      ...
+      },
+      ...
+    },
+    ...
+  },
+  ...
+  },
+
+  "msg": "Decision Log",
+  "query": "data.envoy.authz.allow",
+  "requested_by": "",
+  "result": true,
+  "time": "2020-07-08T20:17:27Z",
+  "timestamp": "2020-07-08T20:17:27.7568234Z",
+  "type": "openpolicyagent.org/decision_logs"
+}
+```
+
+Note the presence of the `authorization` header containing the JWT. As explained in the [Rego Policy](#opa-policy) section, this JWT is decoded using the special-purpose code provided by OPA for dealing with JWTs and then the SPIFFE ID is extracted. As we already know, the SPIFFE ID for the `frontend` service matches the SPIFFE ID defined in the rego policy configured for the OPA Agent. Furthermore, request's path and method also match the rule so the `result` for the decision is `true` and the request is allowed to pass through the filter and reach the `backend` service.
+
+## Testing Invalid Requests
+
+On the other hand, when you connect to the URL for the `frontend-2` service (e.g. `http://35.222.190.182:3002`), the browser only displays the title without any account details. This is because the SPIFFE ID of the `frontend-2` service (`spiffe://example.org/ns/default/sa/default/frontend-2`) does not satisfy the policy for the OPA Agent.
+
+![Frontend-2-no-details][frontend-2-view-no-details]
+
+[frontend-2-view-no-details]: images/frontend-2_view_no_details.png "Frontend-2 view no details account"
+
+After trying to display the `frontend-2` data, you can verify the decision made by OPA using the same `scripts/backend-opa-logs.sh` script as performed in the previous section. A similar log entry is available for the `frontend-2` service but with the `result` equal to `false` due to the SPIFFE ID mismatch.
+
+## Retesting frontend-2 with a New Policy
+
+Let's update the rego policy to match the SPIFFE ID of the `frontend-2` service and test again. There is a Bash script that you can leverage to complete this task. Once executed, it will open the editor defined by your `KUBE_EDITOR`, or `EDITOR` environment variables, or fall back to 'vi' for Linux or 'notepad' for Windows.
```

Comment:
```suggestion
Let's update the Rego policy to match the SPIFFE ID of the `frontend-2` service and test again. There is a Bash script that you can leverage to complete this task. Once executed, it will open the editor defined by your `KUBE_EDITOR`, or `EDITOR` environment variables, or fall back to `vi` for Linux or Notepad for Windows.
```